<a href="https://colab.research.google.com/github/john-pabon/Challenge-Oracle-Decoder/blob/main/ImportandoArchivos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Aula 1 -Leyendo archivos CSV


## Importando archivos CSV

In [304]:
import pandas as pd

In [305]:
#Crear variable para guardar los datos del archivo, despues de leerlo
datos = pd.read_csv('/content/superstore_data.csv')
datos.head()


Id  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0   1826        1970  Graduation       Divorced  84835.0        0         0   
1      1        1961  Graduation         Single  57091.0        0         0   
2  10476        1958  Graduation        Married  67267.0        0         1   
3   1386        1967  Graduation       Together  32474.0        1         1   
4   5371        1989  Graduation         Single  21474.0        1         0   

  Dt_Customer  Recency  MntWines  ...  MntFishProducts  MntSweetProducts  \
0   6/16/2014        0       189  ...              111               189   
1   6/15/2014        0       464  ...                7                 0   
2   5/13/2014        0       134  ...               15                 2   
3   11/5/2014        0        10  ...                0                 0   
4    8/4/2014        0         6  ...               11                 0   

   MntGoldProds  NumDealsPurchases  NumWebPurchases  NumCatalogPurchases  \
0           218                  1                4                    4   
1            37                  1                7                    3   
2            30                  1                3                    2   
3             0                  1                1                    0   
4            34                  2                3                    1   

   NumStorePurchases  NumWebVisitsMonth  Response  Complain  
0                  6                  1         1         0  
1                  7                  5         1         0  
2                  5                  2         0         0  
3                  2                  7         0         0  
4                  2                  7         1         0  

[5 rows x 22 columns]

In [306]:
#Muestra los primero cinco datos
datos.head(5)

Id  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0   1826        1970  Graduation       Divorced  84835.0        0         0   
1      1        1961  Graduation         Single  57091.0        0         0   
2  10476        1958  Graduation        Married  67267.0        0         1   
3   1386        1967  Graduation       Together  32474.0        1         1   
4   5371        1989  Graduation         Single  21474.0        1         0   

  Dt_Customer  Recency  MntWines  ...  MntFishProducts  MntSweetProducts  \
0   6/16/2014        0       189  ...              111               189   
1   6/15/2014        0       464  ...                7                 0   
2   5/13/2014        0       134  ...               15                 2   
3   11/5/2014        0        10  ...                0                 0   
4    8/4/2014        0         6  ...               11                 0   

   MntGoldProds  NumDealsPurchases  NumWebPurchases  NumCatalogPurchases  \
0           218                  1                4                    4   
1            37                  1                7                    3   
2            30                  1                3                    2   
3             0                  1                1                    0   
4            34                  2                3                    1   

   NumStorePurchases  NumWebVisitsMonth  Response  Complain  
0                  6                  1         1         0  
1                  7                  5         1         0  
2                  5                  2         0         0  
3                  2                  7         0         0  
4                  2                  7         1         0  

[5 rows x 22 columns]

In [307]:
# que pasa si el archivo no esta separado por comas sino por punto y comas
datos_punto_coma = pd.read_csv('/content/superstore_data_punto_coma.csv')
datos_punto_coma.head()


Id;Year_Birth;Education;Marital_Status;Income;Kidhome;Teenhome;Dt_Customer;Recency;MntWines;MntFruits;MntMeatProducts;MntFishProducts;MntSweetProducts;MntGoldProds;NumDealsPurchases;NumWebPurchases;NumCatalogPurchases;NumStorePurchases;NumWebVisitsMonth;Response;Complain
0  1826;1970;Graduation;Divorced;84835;0;0;6/16/2...                                                                                                                                                                                                                             
1  1;1961;Graduation;Single;57091;0;0;6/15/2014;0...                                                                                                                                                                                                                             
2  10476;1958;Graduation;Married;67267;0;1;5/13/2...                                                                                                                                                                                                                             
3  1386;1967;Graduation;Together;32474;1;1;11/5/2...                                                                                                                                                                                                                             
4  5371;1989;Graduation;Single;21474;1;0;8/4/2014...

In [308]:
#Como solucionar la lecturaras de datos por punto y coma con la funcion sep
datos_punto_coma = pd.read_csv('/content/superstore_data_punto_coma.csv', sep =';')
datos_punto_coma.head()

Id  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0   1826        1970  Graduation       Divorced  84835.0        0         0   
1      1        1961  Graduation         Single  57091.0        0         0   
2  10476        1958  Graduation        Married  67267.0        0         1   
3   1386        1967  Graduation       Together  32474.0        1         1   
4   5371        1989  Graduation         Single  21474.0        1         0   

  Dt_Customer  Recency  MntWines  ...  MntFishProducts  MntSweetProducts  \
0   6/16/2014        0       189  ...              111               189   
1   6/15/2014        0       464  ...                7                 0   
2   5/13/2014        0       134  ...               15                 2   
3   11/5/2014        0        10  ...                0                 0   
4    8/4/2014        0         6  ...               11                 0   

   MntGoldProds  NumDealsPurchases  NumWebPurchases  NumCatalogPurchases  \
0           218                  1                4                    4   
1            37                  1                7                    3   
2            30                  1                3                    2   
3             0                  1                1                    0   
4            34                  2                3                    1   

   NumStorePurchases  NumWebVisitsMonth  Response  Complain  
0                  6                  1         1         0  
1                  7                  5         1         0  
2                  5                  2         0         0  
3                  2                  7         0         0  
4                  2                  7         1         0  

[5 rows x 22 columns]

In [309]:
#funcion nrows para decir cuantas filas quiere ver
datos_primeras_lineas = pd.read_csv('superstore_data.csv', nrows = 5)
datos_primeras_lineas

Id  Year_Birth   Education Marital_Status  Income  Kidhome  Teenhome  \
0   1826        1970  Graduation       Divorced   84835        0         0   
1      1        1961  Graduation         Single   57091        0         0   
2  10476        1958  Graduation        Married   67267        0         1   
3   1386        1967  Graduation       Together   32474        1         1   
4   5371        1989  Graduation         Single   21474        1         0   

  Dt_Customer  Recency  MntWines  ...  MntFishProducts  MntSweetProducts  \
0   6/16/2014        0       189  ...              111               189   
1   6/15/2014        0       464  ...                7                 0   
2   5/13/2014        0       134  ...               15                 2   
3   11/5/2014        0        10  ...                0                 0   
4    8/4/2014        0         6  ...               11                 0   

   MntGoldProds  NumDealsPurchases  NumWebPurchases  NumCatalogPurchases  \
0           218                  1                4                    4   
1            37                  1                7                    3   
2            30                  1                3                    2   
3             0                  1                1                    0   
4            34                  2                3                    1   

   NumStorePurchases  NumWebVisitsMonth  Response  Complain  
0                  6                  1         1         0  
1                  7                  5         1         0  
2                  5                  2         0         0  
3                  2                  7         0         0  
4                  2                  7         1         0  

[5 rows x 22 columns]

In [310]:
#funcion usecols para seleccionar  que tipo de columnas  quiero ver
datos_seleccion = pd.read_csv('superstore_data.csv', usecols = ['Id','Year_Birth','Income'])
datos_seleccion

Id  Year_Birth   Income
0      1826        1970  84835.0
1         1        1961  57091.0
2     10476        1958  67267.0
3      1386        1967  32474.0
4      5371        1989  21474.0
...     ...         ...      ...
2235  10142        1976  66476.0
2236   5263        1977  31056.0
2237     22        1976  46310.0
2238    528        1978  65819.0
2239   4070        1969  94871.0

[2240 rows x 3 columns]

In [311]:
#funcion usecols para seleccionar la posicion de las  columnas quiero ver
datos_seleccion = pd.read_csv('superstore_data.csv', usecols = [0,1,4])
datos_seleccion

Id  Year_Birth   Income
0      1826        1970  84835.0
1         1        1961  57091.0
2     10476        1958  67267.0
3      1386        1967  32474.0
4      5371        1989  21474.0
...     ...         ...      ...
2235  10142        1976  66476.0
2236   5263        1977  31056.0
2237     22        1976  46310.0
2238    528        1978  65819.0
2239   4070        1969  94871.0

[2240 rows x 3 columns]

In [312]:
#Exportar archivos a CSV
datos_seleccion.to_csv('clientes_mercado.csv')

In [313]:
#Abrir archivo
clientes_mercado = pd.read_csv('clientes_mercado.csv')
clientes_mercado.head()

Unnamed: 0     Id  Year_Birth   Income
0           0   1826        1970  84835.0
1           1      1        1961  57091.0
2           2  10476        1958  67267.0
3           3   1386        1967  32474.0
4           4   5371        1989  21474.0

In [314]:
datos_seleccion.to_csv('datos_mercado.csv', index =False)

In [315]:
#Aparece una columna extra llamada indice y necesitamos removerla
datos_mercado = pd.read_csv('datos_mercado.csv')
datos_mercado

Id  Year_Birth   Income
0      1826        1970  84835.0
1         1        1961  57091.0
2     10476        1958  67267.0
3      1386        1967  32474.0
4      5371        1989  21474.0
...     ...         ...      ...
2235  10142        1976  66476.0
2236   5263        1977  31056.0
2237     22        1976  46310.0
2238    528        1978  65819.0
2239   4070        1969  94871.0

[2240 rows x 3 columns]

Es hora de que pongas a prueba los conocimientos desarrollados durante la clase. Disponemos de un archivo en formato CSV, el cual se obtuvo de la página Información de Salud (TABNET) - DATASUS. Este archivo contiene los gastos hospitalarios públicos de cada una de las Unidades de la Federación Brasileña desde enero de 2008 hasta marzo de 2021.

Su desafío aquí será leer este archivo usando la función read_csv de la biblioteca Pandas. Se deben agregar algunos parámetros para que la lectura se realice correctamente. Entonces, aquí hay algunos consejos:

Asegúrese de que el archivo CSV esté separado por una coma o punto y coma.
La codificación del archivo es ISO-8859-1.
Las primeras tres líneas del archivo se pueden ignorar, ya que el encabezado sólo comienza en la cuarta línea.
Las últimas 9 líneas también se pueden ignorar, ya que son sólo información sobre dónde se tomaron los datos.
Para eliminar las últimas líneas es necesario agregar el parámetro engine='python'.
¿Comenzamos?

In [316]:
#Crear variable para guardar los datos del archivo, despues de leerlo
datos = pd.read_csv('datos_mercado.csv')

# Aula 2 Leyendo archivos Excel

## Importe archivos Excel

In [317]:
import pandas as pd

In [318]:
archivo = 'emisiones_CO2.xlsx'

In [319]:
# leer archivo
pd.read_excel(archivo)

País ISO 3166-1 alpha-3   Año         Total        Carbón  \
0      Afganistán                AFG  1750  0.000000e+00           NaN   
1      Afganistán                AFG  1751  0.000000e+00           NaN   
2      Afganistán                AFG  1752  0.000000e+00           NaN   
3      Afganistán                AFG  1753  0.000000e+00           NaN   
4      Afganistán                AFG  1754  0.000000e+00           NaN   
...           ...                ...   ...           ...           ...   
63099      Global                WLD  2017  3.609674e+10  1.450697e+10   
63100      Global                WLD  2018  3.682651e+10  1.474683e+10   
63101      Global                WLD  2019  3.708256e+10  1.472598e+10   
63102      Global                WLD  2020  3.526409e+10  1.417456e+10   
63103      Global                WLD  2021  3.712385e+10  1.497960e+10   

             Aceite           Gas       Cemento        Quema        Otros  \
0               NaN           NaN           NaN          NaN          NaN   
1               NaN           NaN           NaN          NaN          NaN   
2               NaN           NaN           NaN          NaN          NaN   
3               NaN           NaN           NaN          NaN          NaN   
4               NaN           NaN           NaN          NaN          NaN   
...             ...           ...           ...          ...          ...   
63099  1.224263e+10  7.144928e+09  1.507923e+09  391992176.0  302294047.0   
63100  1.226602e+10  7.529847e+09  1.569218e+09  412115746.0  302478706.0   
63101  1.234565e+10  7.647528e+09  1.617507e+09  439253991.0  306638573.0   
63102  1.119181e+10  7.556290e+09  1.637538e+09  407583673.0  296301685.0   
63103  1.183716e+10  7.921829e+09  1.672592e+09  416525563.0  296145746.0   

       Per Capita  
0             NaN  
1             NaN  
2             NaN  
3             NaN  
4             NaN  
...           ...  
63099   4749682.0  
63100   4792753.0  
63101   4775633.0  
63102   4497423.0  
63103   4693699.0  

[63104 rows x 11 columns]

In [320]:
# convertirlo el metodo dato frame
datos_co2 = pd.read_excel(archivo)

In [321]:
datos_co2.head(5)

País ISO 3166-1 alpha-3   Año  Total  Carbón  Aceite  Gas  Cemento  \
0  Afganistán                AFG  1750    0.0     NaN     NaN  NaN      NaN   
1  Afganistán                AFG  1751    0.0     NaN     NaN  NaN      NaN   
2  Afganistán                AFG  1752    0.0     NaN     NaN  NaN      NaN   
3  Afganistán                AFG  1753    0.0     NaN     NaN  NaN      NaN   
4  Afganistán                AFG  1754    0.0     NaN     NaN  NaN      NaN   

   Quema  Otros  Per Capita  
0    NaN    NaN         NaN  
1    NaN    NaN         NaN  
2    NaN    NaN         NaN  
3    NaN    NaN         NaN  
4    NaN    NaN         NaN

In [322]:
# Ver cual hoja o sheer estoy viendo
pd.ExcelFile(archivo).sheet_names

['emisiones_C02', 'emisiones_percapita', 'fuentes']

In [323]:
# Para ver la segunda hoja o tercera hoja
percapita = pd.read_excel(archivo, sheet_name = 'emisiones_percapita')

In [324]:
# Muestra las primeras cinco lineas
percapita.head()

País ISO 3166-1 alpha-3   Año  Total  Carbón  Aceite  Gas  Cemento  \
0  Afganistán                AFG  1750    NaN     NaN     NaN  NaN      NaN   
1  Afganistán                AFG  1751    NaN     NaN     NaN  NaN      NaN   
2  Afganistán                AFG  1752    NaN     NaN     NaN  NaN      NaN   
3  Afganistán                AFG  1753    NaN     NaN     NaN  NaN      NaN   
4  Afganistán                AFG  1754    NaN     NaN     NaN  NaN      NaN   

   Quema  Otros  
0    NaN    NaN  
1    NaN    NaN  
2    NaN    NaN  
3    NaN    NaN  
4    NaN    NaN

In [325]:
# Muestra los ultimos cinco reistros
percapita.tail()

País ISO 3166-1 alpha-3   Año      Total     Carbón     Aceite  \
63099  Global                WLD  2017  4749682.0  1908857.0  1610910.0   
63100  Global                WLD  2018  4792753.0  1919213.0  1596350.0   
63101  Global                WLD  2019  4775633.0  1896468.0  1589920.0   
63102  Global                WLD  2020  4497423.0  1807760.0  1427353.0   
63103  Global                WLD  2021  4693699.0  1893923.0  1496614.0   

             Gas   Cemento    Quema    Otros  
63099   940144.0  198416.0  51579.0  39776.0  
63100   979965.0  204225.0  53634.0  39366.0  
63101   984878.0  208309.0  56569.0  39490.0  
63102   963695.0  208844.0  51981.0  37789.0  
63103  1001585.0  211472.0  52663.0  37443.0

In [326]:
# Muestra registros aleatorios
percapita.sample(5)

País ISO 3166-1 alpha-3   Año      Total  \
52109                    Sudáfrica                ZAF  1907        NaN   
46674                      Reunión                REU  1912        NaN   
9934   República De África Central                CAF  1892        NaN   
26333                      Irlanda                IRL  1971  7569274.0   
772                        Argelia                DZA  1978  3542025.0   

          Carbón     Aceite        Gas   Cemento      Quema  Otros  
52109        NaN        NaN        NaN       NaN        NaN    NaN  
46674        NaN        NaN        NaN       NaN        NaN    NaN  
9934         NaN        NaN        NaN       NaN        NaN    NaN  
26333  2768918.0  4549465.0        0.0  250892.0        0.0    NaN  
772      32416.0   746612.0  1352962.0   74942.0  1335092.0    NaN

In [327]:
# Verificar las fuentes
fuentes = pd.read_excel(archivo, sheet_name='fuentes')

In [328]:
fuentes.tail()

País ISO 3166-1 alpha-3   Año                                 Total  \
63099  Global                WLD  2017  CDIAC 2022, BP, and Sum of countries   
63100  Global                WLD  2018  CDIAC 2022, BP, and Sum of countries   
63101  Global                WLD  2019  CDIAC 2022, BP, and Sum of countries   
63102  Global                WLD  2020  CDIAC 2022, BP, and Sum of countries   
63103  Global                WLD  2021  CDIAC 2022, BP, and Sum of countries   

                  Carbón             Aceite                Gas        Cemento  \
63099  CDIAC 2022 and BP  CDIAC 2022 and BP  CDIAC 2022 and BP  Andrew cement   
63100  CDIAC 2022 and BP  CDIAC 2022 and BP  CDIAC 2022 and BP  Andrew cement   
63101  CDIAC 2022 and BP  CDIAC 2022 and BP  CDIAC 2022 and BP  Andrew cement   
63102  CDIAC 2022 and BP  CDIAC 2022 and BP  CDIAC 2022 and BP  Andrew cement   
63103  CDIAC 2022 and BP  CDIAC 2022 and BP  CDIAC 2022 and BP  Andrew cement   

                    Quema   Otros  \
63099  CDIAC 2022 and GCP  [NONE]   
63100  CDIAC 2022 and GCP  [NONE]   
63101  CDIAC 2022 and GCP  [NONE]   
63102  CDIAC 2022 and GCP  [NONE]   
63103  CDIAC 2022 and GCP  [NONE]   

                                              Per Capita  
63099  CDIAC 2022, BP, Sum of countries, and UN popul...  
63100  CDIAC 2022, BP, Sum of countries, and UN popul...  
63101  CDIAC 2022, BP, Sum of countries, and UN popul...  
63102  CDIAC 2022, BP, Sum of countries, and UN popul...  
63103  CDIAC 2022, BP, Sum of countries, and UN popul...

In [329]:
# Escoger las primeras cuatro columnas
intervalo = pd.read_excel(archivo, sheet_name ='emisiones_C02', usecols = 'A:E')

In [330]:
# leer la variable
intervalo

País ISO 3166-1 alpha-3   Año         Total        Carbón
0      Afganistán                AFG  1750  0.000000e+00           NaN
1      Afganistán                AFG  1751  0.000000e+00           NaN
2      Afganistán                AFG  1752  0.000000e+00           NaN
3      Afganistán                AFG  1753  0.000000e+00           NaN
4      Afganistán                AFG  1754  0.000000e+00           NaN
...           ...                ...   ...           ...           ...
63099      Global                WLD  2017  3.609674e+10  1.450697e+10
63100      Global                WLD  2018  3.682651e+10  1.474683e+10
63101      Global                WLD  2019  3.708256e+10  1.472598e+10
63102      Global                WLD  2020  3.526409e+10  1.417456e+10
63103      Global                WLD  2021  3.712385e+10  1.497960e+10

[63104 rows x 5 columns]

In [331]:
#Leer las filas
intervalo2 = pd.read_excel(archivo, sheet_name ='emisiones_C02', usecols = 'A:E', nrows =10)

In [332]:
intervalo2

País ISO 3166-1 alpha-3   Año  Total  Carbón
0  Afganistán                AFG  1750      0     NaN
1  Afganistán                AFG  1751      0     NaN
2  Afganistán                AFG  1752      0     NaN
3  Afganistán                AFG  1753      0     NaN
4  Afganistán                AFG  1754      0     NaN
5  Afganistán                AFG  1755      0     NaN
6  Afganistán                AFG  1756      0     NaN
7  Afganistán                AFG  1757      0     NaN
8  Afganistán                AFG  1758      0     NaN
9  Afganistán                AFG  1759      0     NaN

In [333]:
# Exportar archivo a archivo Excel
intervalo.to_excel('co2_percapita.xlsx', index = False)

In [334]:
pd.read_excel('co2_percapita.xlsx')

País ISO 3166-1 alpha-3   Año         Total        Carbón
0      Afganistán                AFG  1750  0.000000e+00           NaN
1      Afganistán                AFG  1751  0.000000e+00           NaN
2      Afganistán                AFG  1752  0.000000e+00           NaN
3      Afganistán                AFG  1753  0.000000e+00           NaN
4      Afganistán                AFG  1754  0.000000e+00           NaN
...           ...                ...   ...           ...           ...
63099      Global                WLD  2017  3.609674e+10  1.450697e+10
63100      Global                WLD  2018  3.682651e+10  1.474683e+10
63101      Global                WLD  2019  3.708256e+10  1.472598e+10
63102      Global                WLD  2020  3.526409e+10  1.417456e+10
63103      Global                WLD  2021  3.712385e+10  1.497960e+10

[63104 rows x 5 columns]

## Leyendo datos de Google Sheets online

### link https://docs.google.com/spreadsheets/d/1ypx_w7PbTVDMySchayKR9ctX2kKNJE521jVahohg094/edit?usp=sharing


In [335]:
sheet_id = '1ypx_w7PbTVDMySchayKR9ctX2kKNJE521jVahohg094'
url = f'https://docs.google.com/spreadsheets/d/{sheet_id}/gviz/tq?tqx=out:csv&sheet'

In [336]:
# leer archivo
url

'https://docs.google.com/spreadsheets/d/1ypx_w7PbTVDMySchayKR9ctX2kKNJE521jVahohg094/gviz/tq?tqx=out:csv&sheet'

In [337]:
datos_co2_sheets = pd.read_csv(url)

####  significa que estás intentando acceder a un archivo de Google Sheets que no está disponible públicamente. Es decir, necesitas permisos para acceder a él.

#### ¿Cómo solucionar este problema?
Tienes dos opciones:
Opción 1: Hacer el archivo público
Abre la hoja de cálculo en Google Sheets.

Haz clic en el botón “Compartir” (arriba a la derecha).

En la ventana emergente, cambia los permisos a:

“Cualquiera con el enlace” puede “Ver”.

Luego usa el siguiente formato para acceder:

Opción 2: Descargar el archivo como CSV y cargarlo localmente
Abre la hoja en Google Sheets.

Ve a Archivo > Descargar > Valores separados por comas (.csv).

En tu notebook Colab, sube el archivo:

#### ¿Qué es esto?
Es una f-string de Python que construye dinámicamente una URL para descargar una hoja de cálculo de Google Sheets como un archivo CSV.

🧩 Parte por parte:
1. f'...'
Es una f-string (formato de cadena). Permite incluir variables dentro de una cadena de texto usando llaves {}.

2. 'https://docs.google.com/spreadsheets/d/'
Esta es la parte inicial de la URL que apunta al sitio de Google Sheets.

3. {sheet_id}
Aquí se inserta el ID del documento de Google Sheets.

Ejemplo: '1ypx_w7PbTVDMySchayKR9ctX2kKNJE521jVahohg094'

Este ID aparece en la URL del documento de Google:

latex
Copy
Edit
https://docs.google.com/spreadsheets/d/1ypx_w7PbTVDMySchayKR9ctX2kKNJE521jVahohg094/edit
                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                                          → Este es el ID
4. /gviz/tq?tqx=out:csv
Esta parte le indica a Google que queremos los datos en formato CSV.

gviz/tq es una interfaz especial de Google que permite consultar datos.

tqx=out:csv dice: "Queremos la salida (out) en formato CSV".

5. &sheet={sheet_name}
Aquí se inserta el nombre de la hoja dentro del archivo.

Ejemplo: si tu hoja se llama "Ventas_2024", esa sería la variable {sheet_name}.

Asegúrate de que coincida exactamente con el nombre que aparece en la pestaña de la hoja.

In [338]:
#solucion 1
import pandas as pd

sheet_id = '1ypx_w7PbTVDMySchayKR9ctX2kKNJE521jVahohg094'
sheet_name = 'NombreDeTuHoja'  # <-- reemplaza con el nombre real de tu hoja

url = f'https://docs.google.com/spreadsheets/d/{sheet_id}/gviz/tq?tqx=out:csv&sheet={sheet_name}'
datos_co2_sheets = pd.read_csv(url)

In [339]:
datos_co2_sheets

País ISO 3166-1 alpha-3   Año           Total          Carbón  \
0      Afganistán                AFG  1750               0             NaN   
1      Afganistán                AFG  1751               0             NaN   
2      Afganistán                AFG  1752               0             NaN   
3      Afganistán                AFG  1753               0             NaN   
4      Afganistán                AFG  1754               0             NaN   
...           ...                ...   ...             ...             ...   
63099      Global                WLD  2017  36,096,739,276  14,506,973,805   
63100      Global                WLD  2018  36,826,506,600  14,746,830,688   
63101      Global                WLD  2019  37,082,558,969  14,725,978,025   
63102      Global                WLD  2020  35,264,085,734  14,174,564,010   
63103      Global                WLD  2021  37,123,850,352  14,979,598,083   

               Aceite            Gas        Cemento        Quema        Otros  \
0                 NaN            NaN            NaN          NaN          NaN   
1                 NaN            NaN            NaN          NaN          NaN   
2                 NaN            NaN            NaN          NaN          NaN   
3                 NaN            NaN            NaN          NaN          NaN   
4                 NaN            NaN            NaN          NaN          NaN   
...               ...            ...            ...          ...          ...   
63099  12,242,627,935  7,144,928,128  1,507,923,185  391,992,176  302,294,047   
63100  12,266,016,285  7,529,846,784  1,569,218,392  412,115,746  302,478,706   
63101  12,345,653,374  7,647,528,220  1,617,506,786  439,253,991  306,638,573   
63102  11,191,808,551  7,556,290,283  1,637,537,532  407,583,673  296,301,685   
63103  11,837,159,116  7,921,829,472  1,672,592,372  416,525,563  296,145,746   

      Per Capita  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN  
...          ...  
63099  4,749,682  
63100  4,792,753  
63101  4,775,633  
63102  4,497,423  
63103  4,693,699  

[63104 rows x 11 columns]

In [340]:
# Quiero ver la segunda hoja
sheet_id = '1ypx_w7PbTVDMySchayKR9ctX2kKNJE521jVahohg094'
sheet_name = 'NombreDeTuHojaemisiones_percapita'  # <-- reemplaza con el nombre real de tu hoja

url_percapita= f'https://docs.google.com/spreadsheets/d/{sheet_id}/gviz/tq?tqx=out:csv&sheet={sheet_name}'


In [342]:
percapita_sheets = pd.read_csv(url_percapita)

In [343]:
percapita_sheets

País ISO 3166-1 alpha-3   Año           Total          Carbón  \
0      Afganistán                AFG  1750               0             NaN   
1      Afganistán                AFG  1751               0             NaN   
2      Afganistán                AFG  1752               0             NaN   
3      Afganistán                AFG  1753               0             NaN   
4      Afganistán                AFG  1754               0             NaN   
...           ...                ...   ...             ...             ...   
63099      Global                WLD  2017  36,096,739,276  14,506,973,805   
63100      Global                WLD  2018  36,826,506,600  14,746,830,688   
63101      Global                WLD  2019  37,082,558,969  14,725,978,025   
63102      Global                WLD  2020  35,264,085,734  14,174,564,010   
63103      Global                WLD  2021  37,123,850,352  14,979,598,083   

               Aceite            Gas        Cemento        Quema        Otros  \
0                 NaN            NaN            NaN          NaN          NaN   
1                 NaN            NaN            NaN          NaN          NaN   
2                 NaN            NaN            NaN          NaN          NaN   
3                 NaN            NaN            NaN          NaN          NaN   
4                 NaN            NaN            NaN          NaN          NaN   
...               ...            ...            ...          ...          ...   
63099  12,242,627,935  7,144,928,128  1,507,923,185  391,992,176  302,294,047   
63100  12,266,016,285  7,529,846,784  1,569,218,392  412,115,746  302,478,706   
63101  12,345,653,374  7,647,528,220  1,617,506,786  439,253,991  306,638,573   
63102  11,191,808,551  7,556,290,283  1,637,537,532  407,583,673  296,301,685   
63103  11,837,159,116  7,921,829,472  1,672,592,372  416,525,563  296,145,746   

      Per Capita  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN  
...          ...  
63099  4,749,682  
63100  4,792,753  
63101  4,775,633  
63102  4,497,423  
63103  4,693,699  

[63104 rows x 11 columns]

In [344]:
# Quiero ver la segunda hoja
sheet_id = '1ypx_w7PbTVDMySchayKR9ctX2kKNJE521jVahohg094'
sheet_name = 'fuentes'  # <-- reemplaza con el nombre real de tu hoja

url_fuentes= f'https://docs.google.com/spreadsheets/d/{sheet_id}/gviz/tq?tqx=out:csv&sheet={sheet_name}'

In [345]:
fuentes_sheets =pd.read_csv(url_fuentes)

In [346]:
fuentes_sheets

País ISO 3166-1 alpha-3   Año  \
0      Afganistán                AFG  1750   
1      Afganistán                AFG  1751   
2      Afganistán                AFG  1752   
3      Afganistán                AFG  1753   
4      Afganistán                AFG  1754   
...           ...                ...   ...   
63099      Global                WLD  2017   
63100      Global                WLD  2018   
63101      Global                WLD  2019   
63102      Global                WLD  2020   
63103      Global                WLD  2021   

                                      Total             Carbón  \
0                                    [NONE]             [NONE]   
1                                    [NONE]             [NONE]   
2                                    [NONE]             [NONE]   
3                                    [NONE]             [NONE]   
4                                    [NONE]             [NONE]   
...                                     ...                ...   
63099  CDIAC 2022, BP, and Sum of countries  CDIAC 2022 and BP   
63100  CDIAC 2022, BP, and Sum of countries  CDIAC 2022 and BP   
63101  CDIAC 2022, BP, and Sum of countries  CDIAC 2022 and BP   
63102  CDIAC 2022, BP, and Sum of countries  CDIAC 2022 and BP   
63103  CDIAC 2022, BP, and Sum of countries  CDIAC 2022 and BP   

                  Aceite                Gas        Cemento  \
0                 [NONE]             [NONE]         [NONE]   
1                 [NONE]             [NONE]         [NONE]   
2                 [NONE]             [NONE]         [NONE]   
3                 [NONE]             [NONE]         [NONE]   
4                 [NONE]             [NONE]         [NONE]   
...                  ...                ...            ...   
63099  CDIAC 2022 and BP  CDIAC 2022 and BP  Andrew cement   
63100  CDIAC 2022 and BP  CDIAC 2022 and BP  Andrew cement   
63101  CDIAC 2022 and BP  CDIAC 2022 and BP  Andrew cement   
63102  CDIAC 2022 and BP  CDIAC 2022 and BP  Andrew cement   
63103  CDIAC 2022 and BP  CDIAC 2022 and BP  Andrew cement   

                    Quema   Otros  \
0                  [NONE]  [NONE]   
1                  [NONE]  [NONE]   
2                  [NONE]  [NONE]   
3                  [NONE]  [NONE]   
4                  [NONE]  [NONE]   
...                   ...     ...   
63099  CDIAC 2022 and GCP  [NONE]   
63100  CDIAC 2022 and GCP  [NONE]   
63101  CDIAC 2022 and GCP  [NONE]   
63102  CDIAC 2022 and GCP  [NONE]   
63103  CDIAC 2022 and GCP  [NONE]   

                                              Per Capita  
0                                                 [NONE]  
1                                                 [NONE]  
2                                                 [NONE]  
3                                                 [NONE]  
4                                                 [NONE]  
...                                                  ...  
63099  CDIAC 2022, BP, Sum of countries, and UN popul...  
63100  CDIAC 2022, BP, Sum of countries, and UN popul...  
63101  CDIAC 2022, BP, Sum of countries, and UN popul...  
63102  CDIAC 2022, BP, Sum of countries, and UN popul...  
63103  CDIAC 2022, BP, Sum of countries, and UN popul...  

[63104 rows x 11 columns]

##3 Desafio


#### Es hora de que pongas a prueba los conocimientos desarrollados durante la clase. Tenemos un link de Google Sheets que contiene datos importantes sobre las emisiones de dióxido de carbono en todo el mundo. El conjunto de datos se obtuvo de Kaggle y consta de las emisiones de CO2 per cápita de todos los países del mundo entre 1990 y 2019.

En este desafío, tu función es leer este link de Google Sheets y luego guardar el DataFrame obtenido en formato CSV. ¿Listo para comenzar?

# Aula 3 Leyendo archivos JSON

## Importar archivos JSON

In [347]:
#Importar librerias
import pandas as pd

In [348]:
#Leer archivo
datos_pacientes = pd.read_json('pacientes.json')

In [349]:
datos_pacientes

ID_paciente Enfermedad_corazon    IMC Fumador Consumo_alcohol AVC  \
0              0                 No  16.60      Sí              No  No   
1              1                 No  20.34      No              No  Sí   
2              2                 No  26.58      Sí              No  No   
3              3                 No  24.21      No              No  No   
4              4                 No  23.71      No              No  No   
..           ...                ...    ...     ...             ...  ..   
995          995                 No  28.29      Sí              No  No   
996          996                 No  25.69      Sí              No  No   
997          997                 Sí  26.99      No              No  No   
998          998                 No  44.29      No              No  No   
999          999                 No  25.83      No              No  No   

     Salud_fisica  Salud_mental Dificultad_caminar Sexo_biologico Rango_edad  \
0               3            30                 No       Femenino      55-59   
1               0             0                 No       Femenino     80 ó +   
2              20            30                 No      Masculino      65-69   
3               0             0                 No       Femenino      75-79   
4              28             0                 Sí       Femenino      40-44   
..            ...           ...                ...            ...        ...   
995            10            20                 No       Femenino      65-69   
996             0             0                 No       Femenino      75-79   
997             1             0                 No      Masculino      70-74   
998            30             0                 Sí       Femenino      70-74   
999             0             0                 No      Masculino      70-74   

       Raza Diabetes Actividad_fisica Salud_general  Horas_sueño Asma  \
0    Blanca       Sí               Sí     Muy buena            5   Sí   
1    Blanca       No               Sí     Muy buena            7   No   
2    Blanca       Sí               Sí     Razonable            8   Sí   
3    Blanca       No               No         Buena            6   No   
4    Blanca       No               Sí     Muy buena            8   No   
..      ...      ...              ...           ...          ...  ...   
995   Negra       Sí               Sí     Razonable            8   No   
996  Blanca       No               Sí     Excelente            8   No   
997  Blanca       No               Sí     Muy buena            8   No   
998  Blanca       Sí               No         Buena            6   Sí   
999   Negra       Sí               No     Excelente            7   No   

    Enfermedad_renal Cancer_piel  
0                 No          Sí  
1                 No          No  
2                 No          No  
3                 No          Sí  
4                 No          No  
..               ...         ...  
995               No          No  
996               No          No  
997               No          No  
998               No          No  
999               No          No  

[1000 rows x 19 columns]

In [350]:
# columnas con informacion anidada
datos_pacientes_2 = pd.read_json('pacientes_2.json')

In [351]:
datos_pacientes_2

Investigación   Año  \
0  Indicadores clave de enfermedades cardíacas  2020   
1  Indicadores clave de enfermedades cardíacas  2020   
2  Indicadores clave de enfermedades cardíacas  2020   

                                           Pacientes  
0  {'ID': '01', 'Rango_edad': '55-59', 'Sexo_biol...  
1  {'ID': '02', 'Rango_edad': '80 ó +', 'Sexo_bio...  
2  {'ID': '03', 'Rango_edad': '65-69', 'Sexo_biol...

## Normailizando archivos JSON

In [352]:
df_normailizado = pd.json_normalize(datos_pacientes_2['Pacientes'])

In [353]:
df_normailizado

ID Rango_edad Sexo_biologico    Raza    IMC Fumador Consumo_alcohol  \
0  01      55-59          Mujer  Blanca  16.60      Sí              No   
1  02     80 ó +          Mujer  Blanca  20.34      No              No   
2  03      65-69      Masculino  Blanca  26.58      Sí              No   

   Salud_física  Salud_mental Dificultad_caminar Actividad_física  \
0             3            30                 No               Sí   
1             0             0                 No               Sí   
2            20            30                 No               Sí   

  Salud_general  Horas_sueño                Problemas_salud  
0     Muy buena            5  [Diabetes, Asma, Cancer_piel]  
1     Muy buena            7                          [AVC]  
2     Muy buena            8               [Diabetes, Asma]

### Escribiendo archivos JSON

In [354]:
df_normailizado.to_json('historico_pacientes_norm.json')

In [355]:
pd.read_json('historico_pacientes_norm.json')


ID Rango_edad Sexo_biologico    Raza    IMC Fumador Consumo_alcohol  \
0   1      55-59          Mujer  Blanca  16.60      Sí              No   
1   2     80 ó +          Mujer  Blanca  20.34      No              No   
2   3      65-69      Masculino  Blanca  26.58      Sí              No   

   Salud_física  Salud_mental Dificultad_caminar Actividad_física  \
0             3            30                 No               Sí   
1             0             0                 No               Sí   
2            20            30                 No               Sí   

  Salud_general  Horas_sueño                Problemas_salud  
0     Muy buena            5  [Diabetes, Asma, Cancer_piel]  
1     Muy buena            7                          [AVC]  
2     Muy buena            8               [Diabetes, Asma]

### Para saber más: obteniendo archivos JSON de APIs

### Para obtener datos en formato JSON podemos utilizar una API (Application Programming Interface). Es una interfaz de programación de aplicaciones que permite la comunicación entre diferentes programas o sistemas.

Hay varios formatos que se pueden utilizar en las API y uno de los más comunes es JSON (JavaScript Object Notation). JSON es un formato ligero para intercambiar información entre sistemas que utiliza la notación basada en objetos de JavaScript.

Tenemos la siguiente situación-problema:

“Una empresa de alimentos saludables tiene problemas con la popularidad de sus productos. A pesar de ofrecer opciones saludables, muchos clientes se muestran reacios a comprar los productos debido a la falta de información sobre los valores nutricionales de las frutas que se utilizan en la producción de alimentos”.

Ante esta situación podemos acceder a una API llamada Fruitvice para recopilar información detallada sobre los valores nutricionales de las frutas que se utilizan en los productos. Con esta información, la empresa podrá actualizar sus empaques y materiales de marketing para resaltar los valores nutricionales de las frutas y así atraer más clientes que buscan opciones saludables. Además, la empresa puede utilizar la información para desarrollar nuevos productos que satisfagan aún más las necesidades nutricionales de su clientela.

Una forma común de acceder a las API es a través de la biblioteca requests. Para eso, también es necesario importar el módulo json:

In [356]:
import requests
import json

In [357]:
#Los datos de API Fruitvice se obtendrán con la función request.get('url').
#Dentro de ella pasamos la dirección API seguida de api/fruit/all para obtener los datos de todas las frutas.
datos_frutas = requests.get('https://fruityvice.com/api/fruit/all')

In [358]:
#Conseguimos recuperar los resultados utilizando la función json.loads().
#Para ello es necesario pasar la variable datos_frutas, previamente creada
#con el atributo text que devuelve el contenido de la respuesta.
resultado = json.loads(datos_frutas.text)

In [359]:
#Luego podemos obtener una vista previa del DataFrame.
tabla =pd.DataFrame(resultado)

In [360]:
tabla

name   id           family             order         genus  \
0            Persimmon   52        Ebenaceae           Rosales     Diospyros   
1           Strawberry    3         Rosaceae           Rosales      Fragaria   
2               Banana    1         Musaceae      Zingiberales          Musa   
3               Tomato    5       Solanaceae         Solanales       Solanum   
4                 Pear    4         Rosaceae           Rosales         Pyrus   
5               Durian   60        Malvaceae          Malvales         Durio   
6           Blackberry   64         Rosaceae           Rosales         Rubus   
7          Lingonberry   65        Ericaceae          Ericales     Vaccinium   
8                 Kiwi   66    Actinidiaceae  Struthioniformes       Apteryx   
9               Lychee   67      Sapindaceae        Sapindales        Litchi   
10           Pineapple   10     Bromeliaceae            Poales        Ananas   
11                 Fig   68         Moraceae           Rosales         Ficus   
12          Gooseberry   69  Grossulariaceae      Saxifragales         Ribes   
13        Passionfruit   70   Passifloraceae      Malpighiales    Passiflora   
14                Plum   71         Rosaceae           Rosales        Prunus   
15              Orange    2         Rutaceae        Sapindales        Citrus   
16          GreenApple   72         Rosaceae           Rosales         Malus   
17           Raspberry   23         Rosaceae           Rosales         Rubus   
18          Watermelon   25    Cucurbitaceae      Cucurbitales     Citrullus   
19               Lemon   26         Rutaceae        Sapindales        Citrus   
20               Mango   27    Anacardiaceae        Sapindales     Mangifera   
21           Blueberry   33         Rosaceae           Rosales      Fragaria   
22               Apple    6         Rosaceae           Rosales         Malus   
23               Guava   37        Myrtaceae          Myrtales       Psidium   
24             Apricot   35         Rosaceae           Rosales        Prunus   
25               Melon   41    Cucurbitaceae     Cucurbitaceae       Cucumis   
26           Tangerine   77         Rutaceae        Sapindales        Citrus   
27            Pitahaya   78        Cactaceae    Caryophyllales     Cactaceae   
28                Lime   44         Rutaceae        Sapindales        Citrus   
29         Pomegranate   79       Lythraceae          Myrtales        Punica   
30         Dragonfruit   80        Cactaceae    Caryophyllales  Selenicereus   
31               Grape   81         Vitaceae           Vitales         Vitis   
32               Morus   82         Moraceae           Rosales         Morus   
33              Feijoa   76        Myrtaceae        Myrtoideae    Sellowiana   
34             Avocado   84        Lauraceae          Laurales        Persea   
35           Kiwifruit   85    Actinidiaceae          Ericales     Actinidia   
36           Cranberry   87        Ericaceae          Ericales     Vaccinium   
37              Cherry    9         Rosaceae           Rosales        Prunus   
38               Peach   86         Rosaceae           Rosales        Prunus   
39           Jackfruit   94         Moraceae           Rosales    Artocarpus   
40        Horned Melon   95    Cucurbitaceae      Cucurbitales       Cucumis   
41            Hazelnut   96       Betulaceae           Fagales       Corylus   
42              Pomelo   98         Rutaceae        Sapindales        Citrus   
43          Mangosteen   99       Clusiaceae      Malpighiales      Garcinia   
44             Pumpkin  100    Cucurbitaceae      Cucurbitales     Cucurbita   
45  Japanese Persimmon  101        Ebenaceae          Ericales     Diospyros   
46              Papaya   42       Caricaceae       Brassicales        Carica   
47              Annona  103       Annonaceae           Rosales       Annonas   
48   Ceylon Gooseberry  104       Salicaceae      Malpighiales      Dovyalis   

                 

# Aula 4 Leyendo paginas en HTML e XML

## Importando paginas web

In [362]:
import pandas as pd

In [363]:
datos_html = pd.read_html('peliculas_wikipedia.html')

In [364]:
datos_html

[       0                      1
 0   1998             100 Movies
 1   1999              100 Stars
 2   2000             100 Laughs
 3   2001            100 Thrills
 4   2002           100 Passions
 5   2003  100 Heroes & Villains
 6   2004              100 Songs
 7   2005       100 Movie Quotes
 8   2005              25 Scores
 9   2006             100 Cheers
 10  2006            25 Musicals
 11  2007   100 Movies (Updated)
 12  2008        AFI's 10 Top 10
 13   vte                    vte,
                             Film  Release year              Director  \
 0                   Citizen Kane          1941          Orson Welles   
 1                     Casablanca          1942        Michael Curtiz   
 2                  The Godfather          1972  Francis Ford Coppola   
 3             Gone with the Wind          1939        Victor Fleming   
 4             Lawrence of Arabia          1962            David Lean   
 ..                           ...           ...                   

In [365]:
type(datos_html)

list

In [366]:
len(datos_html)

3

In [367]:
top_peliculas = datos_html[1]

In [368]:
top_peliculas

Film  Release year              Director  \
0                   Citizen Kane          1941          Orson Welles   
1                     Casablanca          1942        Michael Curtiz   
2                  The Godfather          1972  Francis Ford Coppola   
3             Gone with the Wind          1939        Victor Fleming   
4             Lawrence of Arabia          1962            David Lean   
..                           ...           ...                   ...   
95                 The Searchers          1956             John Ford   
96              Bringing Up Baby          1938          Howard Hawks   
97                    Unforgiven          1992        Clint Eastwood   
98  Guess Who's Coming to Dinner          1967        Stanley Kramer   
99           Yankee Doodle Dandy          1942        Michael Curtiz   

                      Production companies  Rank  
0                       RKO Radio Pictures     1  
1                    Warner Bros. Pictures     2  
2   Paramount Pictures, Alfran Productions     3  
3          Selznick International Pictures     4  
4                         Horizon Pictures     5  
..                                     ...   ...  
95                  C. V. Whitney Pictures    96  
96                      RKO Radio Pictures    97  
97                     The Malpaso Company    98  
98                       Columbia Pictures    99  
99                   Warner Bros. Pictures   100  

[100 rows x 5 columns]

### Escribiendo archivos e HTML

In [369]:
type(top_peliculas)

pandas.core.frame.DataFrame

In [370]:
top_peliculas.to_html('top_peliculas.html', index=False)

In [371]:
#Leer archivo
pd.read_html('top_peliculas.html')

[                            Film  Release year              Director  \
 0                   Citizen Kane          1941          Orson Welles   
 1                     Casablanca          1942        Michael Curtiz   
 2                  The Godfather          1972  Francis Ford Coppola   
 3             Gone with the Wind          1939        Victor Fleming   
 4             Lawrence of Arabia          1962            David Lean   
 ..                           ...           ...                   ...   
 95                 The Searchers          1956             John Ford   
 96              Bringing Up Baby          1938          Howard Hawks   
 97                    Unforgiven          1992        Clint Eastwood   
 98  Guess Who's Coming to Dinner          1967        Stanley Kramer   
 99           Yankee Doodle Dandy          1942        Michael Curtiz   
 
                       Production companies  Rank  
 0                       RKO Radio Pictures     1  
 1                

In [372]:
top_peliculas.to_csv('top_peliculas_1998.csv', index=False)

In [373]:
datos = pd.read_csv('top_peliculas_1998.csv')

In [374]:
datos

Film  Release year              Director  \
0                   Citizen Kane          1941          Orson Welles   
1                     Casablanca          1942        Michael Curtiz   
2                  The Godfather          1972  Francis Ford Coppola   
3             Gone with the Wind          1939        Victor Fleming   
4             Lawrence of Arabia          1962            David Lean   
..                           ...           ...                   ...   
95                 The Searchers          1956             John Ford   
96              Bringing Up Baby          1938          Howard Hawks   
97                    Unforgiven          1992        Clint Eastwood   
98  Guess Who's Coming to Dinner          1967        Stanley Kramer   
99           Yankee Doodle Dandy          1942        Michael Curtiz   

                      Production companies  Rank  
0                       RKO Radio Pictures     1  
1                    Warner Bros. Pictures     2  
2   Paramount Pictures, Alfran Productions     3  
3          Selznick International Pictures     4  
4                         Horizon Pictures     5  
..                                     ...   ...  
95                  C. V. Whitney Pictures    96  
96                      RKO Radio Pictures    97  
97                     The Malpaso Company    98  
98                       Columbia Pictures    99  
99                   Warner Bros. Pictures   100  

[100 rows x 5 columns]

### Leyendo / escribiendo archivos XML

In [375]:
#Leer archivo y convertirlo en dataframe
datos_imb = pd.read_xml('imdb_top_1000.xml')

In [376]:
datos_imb.head(3)

index                                        Poster_Link  \
0      0  https://m.media-amazon.com/images/M/MV5BMDFkYT...   
1      1  https://m.media-amazon.com/images/M/MV5BM2MyNj...   
2      2  https://m.media-amazon.com/images/M/MV5BMTMxNT...   

               Series_Title Released_Year Certificate  Runtime  \
0  The Shawshank Redemption          1994           A  142 min   
1             The Godfather          1972           A  175 min   
2           The Dark Knight          2008          UA  152 min   

                  Genre  IMDB_Rating  \
0                 Drama          9.3   
1          Crime, Drama          9.2   
2  Action, Crime, Drama          9.0   

                                            Overview  Meta_score  \
0  Two imprisoned men bond over a number of years...        80.0   
1  An organized crime dynasty's aging patriarch t...       100.0   
2  When the menace known as the Joker wreaks havo...        84.0   

               Director           Star1           Star2          Star3  \
0        Frank Darabont     Tim Robbins  Morgan Freeman     Bob Gunton   
1  Francis Ford Coppola   Marlon Brando       Al Pacino     James Caan   
2     Christopher Nolan  Christian Bale    Heath Ledger  Aaron Eckhart   

            Star4  No_of_Votes        Gross  
0  William Sadler      2343110   28,341,469  
1    Diane Keaton      1620367  134,966,411  
2   Michael Caine      2303232  534,858,444

In [377]:
# Exportar archivo
datos_imb.to_xml('peliculas_imdb.xml')

### Desafío: leer una tabla de una página web

### Es el momento de poner a prueba los conocimientos desarrollados durante la clase.

Vanessa es una científica de datos que está haciendo algunos análisis con datos ambientales. Está desarrollando un proyecto para evaluar el impacto ambiental de las actividades humanas en diferentes países del mundo, pero para ello necesita estimaciones de población de esos países. Al buscar en Internet, encontró una tabla de estimaciones de población en un artículo de la página de Wikipedia.

Al igual que Vanessa, tu desafío es obtener un DataFrame de la tabla que contiene la información del número de habitantes de cada país.

In [378]:
#Leer archivo y convertirlo en dataframe
import pandas as pd

# Leer las tablas desde el archivo HTML/XML
df_list = pd.read_html("Anexo_Países.xml")

# Mostrar la primera tabla encontrada (la que contiene los países y sus datos)
df = df_list[0]
df.head()


N.º País (o territorio dependiente)  \
0   1                           India   
1   2                      China[12]​   
2   3                  Estados Unidos   
3   4                       Indonesia   
4   5                        Pakistán   

  Proyección exponencial de la población al 1/7/2025[7]​ Total mun- dial (%)  \
0                                      1 417 492 000                    1754   
1                                      1 407 934 000                    1742   
2                                        342 181 000                     423   
3                                        285 783 000                     354   
4                                        256 204 000                     317   

  Cambio medio anual (%)[8]​ Cambio absoluto anual promedio  \
0                        090                     12 692 000   
1                       -005                     -1 390 000   
2                        061                      2 084 000   
3                        117                      3 337 000   
4                        256                      6 567 000   

  Cambio medio abs. total anual (%)  \
0                              1524   
1                                 -   
2                               250   
3                               401   
4                               788   

  Años para even- tual du- pli- ca- ción[9]​  \
0                                         78   
1                                          -   
2                                        114   
3                                         60   
4                                         27   

  Censo más reciente, última estimación oficial, proyección de la ONU o reloj de población nacional  \
0                                      1 417 492 000                                                  
1                                      1 408 280 000                                                  
2                                        341 820 000                                                  
3                                        282 477 584                                                  
4                                        241 499 431                                                  

  Fecha de esta última cifra de población en particular (d/mm/aaaa)[10]​  \
0                                          1/07/2025                       
1                                         31/12/2024                       
2                                         23/05/2025                       
3                                         30/06/2024                       
4                                          1/03/2023                       

  Tipo[11]​  \
0         A   
1         E   
2         R   
3         E   
4         C   

  Enlace o hipervínculo (usualmente de tipo oficial) de esta última cifra de población  
0                                    web.archive.org                                    
1                                   www.stats.gov.cn                                    
2                                     www.census.gov                                    
3                  dispendukcapil.acehtimurkab.go.id                                    
4                                     www.pbs.gov.pk

## Aula 5 -Leyendo banco de datos



### Creando un banco de datos local

In [379]:
# importar biblioteca SQL
import sqlalchemy

In [380]:
# Importar metodos de sql
from sqlalchemy import create_engine, MetaData, Table, inspect, text

In [381]:
# Crear motor para soportar tablas
engine = create_engine('sqlite:///:memory')

### Escribiendo en un banco de datos

In [382]:
import pandas as pd

In [383]:
archivo = 'clientes_banco.csv'

In [384]:
# Hay que pasar el archivo a un dataframe para luego crear la base de datos
datos = pd.read_csv(archivo)

In [385]:
datos

ID_Cliente  Edad     Grado_estudio      Estado_civil  Tamaño_familia  \
0          5008804    32    Nivel superior  Relación-estable               2   
1          5008805    32    Nivel superior  Relación-estable               2   
2          5008806    58  Nivel intermedio            Casado               2   
3          5008808    52  Nivel intermedio           Soltero               1   
4          5008809    52  Nivel intermedio           Soltero               1   
...            ...   ...               ...               ...             ...   
438458     6840104    62  Nivel intermedio        Divorciado               1   
438459     6840222    43  Nivel intermedio           Soltero               1   
438460     6841878    22    Nivel superior           Soltero               1   
438461     6842765    59  Nivel intermedio            Casado               2   
438462     6842885    51  Nivel intermedio            Casado               2   

       Categoria_de_renta           Ocupacion  Años_empleado  \
0                Empleado                Otro             12   
1                Empleado                Otro             12   
2                Empleado           Seguridad              3   
3        Business Partner              Ventas              8   
4        Business Partner              Ventas              8   
...                   ...                 ...            ...   
438458        Pensionista                Otro              0   
438459           Empleado  Construcción Civil              8   
438460   Business Partner              Ventas              1   
438461        Pensionista                Otro              0   
438462           Empleado              Ventas              3   

        Rendimiento_anual  Tiene_carro                  Vivienda  
0                427500.0            1    Departamento alquilado  
1                427500.0            1    Departamento alquilado  
2                112500.0            1  Casa/Departamento propio  
3                270000.0            0  Casa/Departamento propio  
4                270000.0            0  Casa/Departamento propio  
...                   ...          ...                       ...  
438458           135000.0            0  Casa/Departamento propio  
438459           103500.0            0  Casa/Departamento propio  
438460            54000.0            0       Vive con los padres  
438461            72000.0            0  Casa/Departamento propio  
438462           121500.0            0  Casa/Departamento propio  

[438463 rows x 11 columns]

In [386]:
datos.head(5)

ID_Cliente  Edad     Grado_estudio      Estado_civil  Tamaño_familia  \
0     5008804    32    Nivel superior  Relación-estable               2   
1     5008805    32    Nivel superior  Relación-estable               2   
2     5008806    58  Nivel intermedio            Casado               2   
3     5008808    52  Nivel intermedio           Soltero               1   
4     5008809    52  Nivel intermedio           Soltero               1   

  Categoria_de_renta  Ocupacion  Años_empleado  Rendimiento_anual  \
0           Empleado       Otro             12           427500.0   
1           Empleado       Otro             12           427500.0   
2           Empleado  Seguridad              3           112500.0   
3   Business Partner     Ventas              8           270000.0   
4   Business Partner     Ventas              8           270000.0   

   Tiene_carro                  Vivienda  
0            1    Departamento alquilado  
1            1    Departamento alquilado  
2            1  Casa/Departamento propio  
3            0  Casa/Departamento propio  
4            0  Casa/Departamento propio

In [387]:
#Exportar a una tabla SQL, nombre a la tabla =cliente, segundo parametro el motor donde se encentra la base = engine, quitar indice como nueva columna
datos.to_sql('clientes', engine, index = False)

ValueError: Table 'clientes' already exists.

In [388]:
# como saber si creo una tabla dentro engine
inspector = inspect(engine)

In [389]:
print(inspector.get_table_names())

['clientes', 'empleados']


#### Para saber más: cláusulas SQL
 Siguiente pregunta

SQL (Structured Query Language - Lenguaje de consulta estructurado) es un lenguaje de consulta utilizado en bases de datos relacionales para insertar, actualizar, consultar y administrar datos.

Para hacer todo esto, existen cláusulas SQL, que son componentes fundamentales de las sentencias SQL, permitiendo especificar detalles sobre cómo se va a realizar la consulta u operación de la base de datos. Las cláusulas se utilizan para filtrar, ordenar, agrupar y limitar los resultados de la consulta.

Las sentencias SQL pueden estar compuestas por una o más cláusulas que proporcionan información adicional sobre lo que se supone que debe hacer la consulta. Las cláusulas más comunes son:

* SELECT: especifica qué columnas deben seleccionarse en la consulta.
* FROM: Especifica las tablas de la base de datos que se consultarán.
* WHERE: Filtra los resultados de la consulta según una o más condiciones especificadas.
* ORDER BY: Ordena los resultados de la consulta en orden ascendente o descendente según una o más columnas.
* GROUP BY: agrupa los resultados de la consulta basados en una o más columnas.
* LIMIT: limita el número de filas devueltas por los resultados de la consulta.

## Leyendo una consulta SQL

In [390]:
# Realizar consultas
query = 'SELECT * FROM clientes WHERE categoria_de_renta = "Empleado"'

In [391]:
#consulta o como se lee
empleados= pd.read_sql(sql = text(query), con = engine.connect())

In [392]:
type(empleados)

pandas.core.frame.DataFrame

In [393]:
empleados

ID_Cliente  Edad     Grado_estudio      Estado_civil  Tamaño_familia  \
0          5008804    32    Nivel superior  Relación-estable               2   
1          5008805    32    Nivel superior  Relación-estable               2   
2          5008806    58  Nivel intermedio            Casado               2   
3          5008815    46    Nivel superior            Casado               2   
4          5112956    46    Nivel superior            Casado               2   
...            ...   ...               ...               ...             ...   
226054     6837905    43  Nivel intermedio            Casado               3   
226055     6837906    43  Nivel intermedio            Casado               3   
226056     6839936    34  Nivel intermedio            Casado               3   
226057     6840222    43  Nivel intermedio           Soltero               1   
226058     6842885    51  Nivel intermedio            Casado               2   

       Categoria_de_renta           Ocupacion  Años_empleado  \
0                Empleado                Otro             12   
1                Empleado                Otro             12   
2                Empleado           Seguridad              3   
3                Empleado        Contabilidad              2   
4                Empleado        Contabilidad              2   
...                   ...                 ...            ...   
226054           Empleado                Otro              7   
226055           Empleado                Otro              7   
226056           Empleado  Construcción Civil              5   
226057           Empleado  Construcción Civil              8   
226058           Empleado              Ventas              3   

        Rendimiento_anual  Tiene_carro                  Vivienda  
0                427500.0            1    Departamento alquilado  
1                427500.0            1    Departamento alquilado  
2                112500.0            1  Casa/Departamento propio  
3                270000.0            1  Casa/Departamento propio  
4                270000.0            1  Casa/Departamento propio  
...                   ...          ...                       ...  
226054           355050.0            1  Casa/Departamento propio  
226055           355050.0            1  Casa/Departamento propio  
226056           135000.0            1  Casa/Departamento propio  
226057           103500.0            0  Casa/Departamento propio  
226058           121500.0            0  Casa/Departamento propio  

[226059 rows x 11 columns]

In [395]:
#Leer tabla completa
pd.read_sql_table('empleados', con = engine.connect())

ID_Cliente  Edad     Grado_estudio      Estado_civil  Tamaño_familia  \
0          5008804    32    Nivel superior  Relación-estable               2   
1          5008805    32    Nivel superior  Relación-estable               2   
2          5008806    58  Nivel intermedio            Casado               2   
3          5008815    46    Nivel superior            Casado               2   
4          5112956    46    Nivel superior            Casado               2   
...            ...   ...               ...               ...             ...   
226054     6837905    43  Nivel intermedio            Casado               3   
226055     6837906    43  Nivel intermedio            Casado               3   
226056     6839936    34  Nivel intermedio            Casado               3   
226057     6840222    43  Nivel intermedio           Soltero               1   
226058     6842885    51  Nivel intermedio            Casado               2   

       Categoria_de_renta           Ocupacion  Años_empleado  \
0                Empleado                Otro             12   
1                Empleado                Otro             12   
2                Empleado           Seguridad              3   
3                Empleado        Contabilidad              2   
4                Empleado        Contabilidad              2   
...                   ...                 ...            ...   
226054           Empleado                Otro              7   
226055           Empleado                Otro              7   
226056           Empleado  Construcción Civil              5   
226057           Empleado  Construcción Civil              8   
226058           Empleado              Ventas              3   

        Rendimiento_anual  Tiene_carro                  Vivienda  
0                427500.0            1    Departamento alquilado  
1                427500.0            1    Departamento alquilado  
2                112500.0            1  Casa/Departamento propio  
3                270000.0            1  Casa/Departamento propio  
4                270000.0            1  Casa/Departamento propio  
...                   ...          ...                       ...  
226054           355050.0            1  Casa/Departamento propio  
226055           355050.0            1  Casa/Departamento propio  
226056           135000.0            1  Casa/Departamento propio  
226057           103500.0            0  Casa/Departamento propio  
226058           121500.0            0  Casa/Departamento propio  

[226059 rows x 11 columns]

In [288]:
# Leer filtrando algunas columnas
pd.read_sql_table('empleados', con = engine.connect(), columns =['ID_Cliente', 'Grado_estudio', 'Rendimiento_anual'] )

ID_Cliente     Grado_estudio  Rendimiento_anual
0          5008804    Nivel superior           427500.0
1          5008805    Nivel superior           427500.0
2          5008806  Nivel intermedio           112500.0
3          5008815    Nivel superior           270000.0
4          5112956    Nivel superior           270000.0
...            ...               ...                ...
226054     6837905  Nivel intermedio           355050.0
226055     6837906  Nivel intermedio           355050.0
226056     6839936  Nivel intermedio           135000.0
226057     6840222  Nivel intermedio           103500.0
226058     6842885  Nivel intermedio           121500.0

[226059 rows x 3 columns]

### Actualizando un banco de datos

In [289]:
query = 'SELECT * FROM clientes'

In [290]:
pd.read_sql(sql=text(query), con = engine.connect())

ID_Cliente  Edad     Grado_estudio      Estado_civil  Tamaño_familia  \
0          5008804    32    Nivel superior  Relación-estable               2   
1          5008805    32    Nivel superior  Relación-estable               2   
2          5008806    58  Nivel intermedio            Casado               2   
3          5008808    52  Nivel intermedio           Soltero               1   
4          5008809    52  Nivel intermedio           Soltero               1   
...            ...   ...               ...               ...             ...   
438458     6840104    62  Nivel intermedio        Divorciado               1   
438459     6840222    43  Nivel intermedio           Soltero               1   
438460     6841878    22    Nivel superior           Soltero               1   
438461     6842765    59  Nivel intermedio            Casado               2   
438462     6842885    51  Nivel intermedio            Casado               2   

       Categoria_de_renta           Ocupacion  Años_empleado  \
0                Empleado                Otro             12   
1                Empleado                Otro             12   
2                Empleado           Seguridad              3   
3        Business Partner              Ventas              8   
4        Business Partner              Ventas              8   
...                   ...                 ...            ...   
438458        Pensionista                Otro              0   
438459           Empleado  Construcción Civil              8   
438460   Business Partner              Ventas              1   
438461        Pensionista                Otro              0   
438462           Empleado              Ventas              3   

        Rendimiento_anual  Tiene_carro                  Vivienda  
0                427500.0            1    Departamento alquilado  
1                427500.0            1    Departamento alquilado  
2                112500.0            1  Casa/Departamento propio  
3                270000.0            0  Casa/Departamento propio  
4                270000.0            0  Casa/Departamento propio  
...                   ...          ...                       ...  
438458           135000.0            0  Casa/Departamento propio  
438459           103500.0            0  Casa/Departamento propio  
438460            54000.0            0       Vive con los padres  
438461            72000.0            0  Casa/Departamento propio  
438462           121500.0            0  Casa/Departamento propio  

[438463 rows x 11 columns]

In [291]:
# Como borrar un cliente
from sqlalchemy.exc import SQLAlchemyError
query = "DELETE FROM  clientes WHERE ID_Cliente = 5008804"
try:
  r_set = engine.connect().execute(text(query))
except SQLAlchemyError as e:
  print(e)
else:
  print("#Registros borrados: ", r_set.rowcount)

(sqlite3.OperationalError) database is locked
[SQL: DELETE FROM  clientes WHERE ID_Cliente = 5008804]
(Background on this error at: https://sqlalche.me/e/20/e3q8)


In [292]:
#Vamos a confirmar
pd.read_sql_table('clientes', con = engine.connect())

ID_Cliente  Edad     Grado_estudio      Estado_civil  Tamaño_familia  \
0          5008804    32    Nivel superior  Relación-estable               2   
1          5008805    32    Nivel superior  Relación-estable               2   
2          5008806    58  Nivel intermedio            Casado               2   
3          5008808    52  Nivel intermedio           Soltero               1   
4          5008809    52  Nivel intermedio           Soltero               1   
...            ...   ...               ...               ...             ...   
438458     6840104    62  Nivel intermedio        Divorciado               1   
438459     6840222    43  Nivel intermedio           Soltero               1   
438460     6841878    22    Nivel superior           Soltero               1   
438461     6842765    59  Nivel intermedio            Casado               2   
438462     6842885    51  Nivel intermedio            Casado               2   

       Categoria_de_renta           Ocupacion  Años_empleado  \
0                Empleado                Otro             12   
1                Empleado                Otro             12   
2                Empleado           Seguridad              3   
3        Business Partner              Ventas              8   
4        Business Partner              Ventas              8   
...                   ...                 ...            ...   
438458        Pensionista                Otro              0   
438459           Empleado  Construcción Civil              8   
438460   Business Partner              Ventas              1   
438461        Pensionista                Otro              0   
438462           Empleado              Ventas              3   

        Rendimiento_anual  Tiene_carro                  Vivienda  
0                427500.0            1    Departamento alquilado  
1                427500.0            1    Departamento alquilado  
2                112500.0            1  Casa/Departamento propio  
3                270000.0            0  Casa/Departamento propio  
4                270000.0            0  Casa/Departamento propio  
...                   ...          ...                       ...  
438458           135000.0            0  Casa/Departamento propio  
438459           103500.0            0  Casa/Departamento propio  
438460            54000.0            0       Vive con los padres  
438461            72000.0            0  Casa/Departamento propio  
438462           121500.0            0  Casa/Departamento propio  

[438463 rows x 11 columns]

In [294]:
#Actualizar registros
from sqlalchemy.exc import SQLAlchemyError
query = "UPDATE FROM  clientes SET Grado_estudio = 'Nivel superior' WHERE ID_Cliente = 5008808"
try:
  r_set = engine.connect().execute(text(query))
except SQLAlchemyError as e:
  print(e)
else:
  print("#Registros actualizados: ", r_set.rowcount)


(sqlite3.OperationalError) near "FROM": syntax error
[SQL: UPDATE FROM  clientes SET Grado_estudio = 'Nivel superior' WHERE ID_Cliente = 5008808]
(Background on this error at: https://sqlalche.me/e/20/e3q8)


In [296]:
inspector = inspect(engine)
print(inspector.get_table_names())

['clientes', 'empleados']


### 09
Desafío: manipulando una base de datos
 Siguiente pregunta

Es el momento de poner a prueba tus conocimientos adquiridos durante la clase.

Usted es responsable de crear una base de datos local de clientes para una institución financiera. Disponemos del archivo CSV con los datos de los clientes.

Tu misión es:

1. Crear la base de datos local con la biblioteca SQLAlchemy.
2. Escribir los datos del archivo CSV en esta base de datos local.
3. Realizar tres actualizaciones en la base de datos:

* Actualizar el registro del cliente ID 6840104 cuyo rendimiento anual cambió a 300000.
* Eliminar el registro de cliente ID 5008809, ya que esta persona ya no tiene cuenta en la institución financiera.
* Crear un nuevo registro de cliente siguiendo las especificaciones a continuación:
- ID_cliente: 6850985
- Edad: 33
- Grado_estudio: Doctorado
- Estado_civil: Soltero
- Tamaño_familia: 1
- Categoria_de_renta: Empleado
- Ocupacion: TI
- Años_empleado: 2
- Rendimiento_anual: 290000
- Tiene_carro: 0
- Vivienda: Casa/Departamento propio

Consejo importante: Para agregar un nuevo cliente a la tabla, use la cláusula INSERT INTO seguida del nombre de la tabla y luego especifique los nombres de las columnas de la tabla entre paréntesis. Utilice la cláusula VALUES y luego pase entre paréntesis los nuevos valores para este nuevo registro. Asegúrese de que los valores estén en el orden correcto y en el formato adecuado para cada columna.